File: LeetCode/19.Remove_Nth_Node_From_End_of_List/Mario/19.Remove_Nth_Node_From_End_of_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode node = dummy;

        int len = 0;
        while (node.next != null) {
            len += 1;
            node = node.next;
        }

        if (len < n) {
            return head;
        } else if (len == n) {
            return head.next;
        }

        ListNode fast = dummy, slow = dummy;

        int i = 0;
        while (i < n && fast.next != null) {
            i += 1;
            fast = fast.next;
        }

        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;

        return dummy.next;
    }
}

File: LeetCode/19.Remove_Nth_Node_From_End_of_List/Mario/19.Remove_Nth_Node_From_End_of_List.py
#!/usr/bin/env python3


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # two pointer,
        dummy = ListNode(-1)
        dummy.next = head
        count = 0

        def reverse(pre):
            nonlocal count
            tmp = None
            cur = pre.next
            while cur:
                count += 1
                nxt = cur.next
                cur.next = tmp
                tmp = cur
                cur = nxt
            pre.next = tmp

        reverse(dummy)

        # no need to remove node
        if count < n:
            reverse(dummy)
            return head

        # remove nth node
        pre, cur = dummy, dummy.next
        count = 1

        while count <= n:
            if count == n:
                pre.next = cur.next
            else:
                cur = cur.next
                pre = pre.next
            count += 1

        reverse(dummy)

        return dummy.next
